source/opt/fix_storage_class.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "fix_storage_class.h"
+
+#include "source/opt/instruction.h"
+#include "source/opt/ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status FixStorageClass::Process() {
+  bool modified = false;
+
+  get_module()->ForEachInst([this, &modified](Instruction* inst) {
+    if (FixInstruction(inst)) {
+      modified = true;
+    }
+  });
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+bool FixStorageClass::FixInstruction(Instruction* inst) {
+  switch (inst->opcode()) {
+    case SpvOpAccessChain:
+    case SpvOpPtrAccessChain:
+    case SpvOpInBoundsAccessChain:
+    case SpvOpCopyObject:
+    case SpvOpPhi:
```

Comment:
`SpvOpPhi` can have multiple arguments that are variables.
Fixing only the first argument may not solve the problem?

Maybe we should replace it with the following code:
```
case SpvOpPhi: {
bool modified = false;
for (size_t i = 0; i < inst->operands().size(); ++i)
  modified ||= FixAddressInstruction(inst, i);
return modified;
```

Reply:
This code fixes the result (not the operands) to match the operands.  I wanted to keep it simple and only have a single forward flow: The operands determine the result.  The idea would be that the operands are changed by looking at the instructions that define them.  This will be clearer when I add more tests.